### PR TITLE
SANSILLAutoProcess multiple transmissions

### LINKED ISSUE #30184: Support for multiple transmissions in SANSILLAutoProcess
### Expected behavior

User can provide one or multiple transmissions for data processing in SANSILLAutoProcess.

### Actual behavior

Currently, only one (or none) transmissions of each type (sample transmission, container transmission, etc.) can be provided for to SANSILLAutoProcess.

### Steps to reproduce the behavior

Provide more than one transmission to SANSILLAutoProcess, e.g. using the example from documentation, modify the `sample_tr` to contain multiple numors:
```
beams = '2866,2867+2868,2878'
containers = '2888+2971,2884+2960,2880+2949'
container_tr = '2870+2954'
beam_tr = '2867+2868'
samples = ['2889,2885,2881',
           '2887,2883,2879',
           '3187,3177,3167']
sample_tr = '2871, 2869, 3172'
thick = [0.1, 0.2, 0.2]

# reduce samples
for i in range(len(samples)):
    SANSILLAutoProcess(
        SampleRuns=samples[i],
        BeamRuns=beams,
        ContainerRuns=containers,
        MaskFiles='mask1.nxs,mask2.nxs,mask3.nxs',
        SensitivityMaps='sens-lamp.nxs',
        SampleTransmissionRuns=sample_tr,
        ContainerTransmissionRuns=container_tr,
        TransmissionBeamRuns=beam_tr,
        SampleThickness=thick[i],
        OutputWorkspace='iq_s' + str(i + 1)
    )
```

Observe the issues from the input validation step.

### Platforms affected
all

### PR DESCRIPTION
**Pull request by darominski**
**Description of work.**

Added support for multiple transmissions to be processed by SANSILLAutoProcess. The dimensionality of the provided transmission input must be either one or consistent with the dimensionality of the relevant sample (one from absorber, beam, container, sample). In case one transmission numor is provided, that transmission workspace will be applied to all provided numors. When more than one transmission is provided (and is consistent with the number of relevant sample numors), each transmission will be applied to data in the same order as sample numors are processed.

**To test:**
1. Make sure you have SystemTest data with ILL/D11 directory in your user data path.
2. Run the following through the mantidpython or workbench:

```
beams = '1020,947,1088'
containers = '1023,973,1003'
container_tr = '1023,988,988'
beam_tr = '1020,1119,1119'
samples = '1025,975,1005'
sample_tr = '1204,990,990'
thick = 0.1

#mask edges first:
MaskBTP(Instrument='D11', Tube='1-3,253-256')
RenameWorkspace(InputWorkspace='D11MaskBTP', OutputWorkspace='mask_vertical')
MaskBTP(Instrument='D11', Pixel='1-3,253-256')
Plus(LHSWorkspace='mask_vertical', RHSWorkspace='D11MaskBTP', OutputWorkspace='edge_masks')
# then, the edges mask can be used as a default mask for all distances and wavelengths
MaskBTP(Instrument='D11', Tube='116-139', Pixel='90-116')
RenameWorkspace(InputWorkspace='D11MaskBTP', OutputWorkspace='mask_39m_10A')
MaskBTP(Instrument='D11', Tube='115-140', Pixel='115-140')
RenameWorkspace(InputWorkspace='D11MaskBTP', OutputWorkspace='mask_8m_4_6A')
MaskBTP(Instrument='D11', Tube='105-145', Pixel='105-145')
RenameWorkspace(InputWorkspace='D11MaskBTP', OutputWorkspace='mask_1m_4_6A')

# reduce samples
# this also tests that already loaded workspace can be passed instead of a file
LoadNexusProcessed(Filename='sens-lamp.nxs', OutputWorkspace='sens-lamp')
SANSILLAutoProcess(
            SampleRuns=samples,
            BeamRuns=beams,
            ContainerRuns=containers,
            DefaultMaskFile='edge_masks',
            MaskFiles='mask_39m_10A,mask_8m_4_6A,mask_1m_4_6A',
            SensitivityMaps='sens-lamp',
            SampleTransmissionRuns=sample_tr,
            ContainerTransmissionRuns=container_tr,
            TransmissionBeamRuns=beam_tr,
            SampleThickness=thick,
            CalculateResolution='MildnerCarpenter',
            OutputWorkspace='iq_mult_wavelengths',
            BeamRadius='0.05',
            TransmissionBeamRadius=0.05
)
```
3. There should be no issues from the input validation or the processing.
4. Try changing the number of numors in the sample_tr to more or less than 3 entries, observe the input validation feedback: the number of transmission inputs should be either consistent with the number of sample runs or equal to one. 

Fixes #30184

Release notes added.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by gvardany** [comment]
Can you make the example in PR description to actually use 2 different transmissions for the same sample?

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcess.py [hunk 1]
```diff
@@ -449,55 +443,84 @@ def PyExec(self):
...
+            [process_transmission_absorber, transmission_absorber_name] = \
+                needs_processing(transmission, 'Absorber')
+            absorber_transmission_names.append(transmission_absorber_name)
+            self.progress.report('Processing transmission absorber')
+            if process_transmission_absorber:
+                SANSILLReduction(Run=transmission,
+                                 ProcessAs='Absorber',
+                                 NormaliseBy=self.normalise,
+                                 OutputWorkspace=transmission_absorber_name)
+        for transmission_no, transmission in enumerate(self.btransmission.split(',')):
```
Comment (gvardany) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcess.py:460]:
```suggestion
        for beam_no, beam in enumerate(self.btransmission.split(',')):
```

File: Testing/SystemTests/tests/framework/SANSILLAutoProcessTest.py [hunk 2]
```diff
@@ -172,6 +172,59 @@ def runTest(self):
...
+        self.tolerance_is_rel_err = True
+        return ['iq_mult_wavelengths', 'D11_AutoProcess_Multiple_Tr_Reference.nxs']
+
+    def runTest(self):
+        beams = '950,1020,947'
+        containers = '958,1023,973'
+        container_tr = '988,1023,988'
+        beam_tr = '947,1020,947'
+        samples = '960,1024,975'
+        sample_tr = '990,1024,990'
```
Comment (gvardany) [inline Testing/SystemTests/tests/framework/SANSILLAutoProcessTest.py:205]: Just to confirm, is it really the middle distance that is at different wavelength? I though it's the first or the last?

**Review by gvardany — COMMENTED** [review]
Few comment inline:

File: Testing/SystemTests/tests/framework/SANSILLAutoProcessTest.py (hunk 2 above)
Comment (darominski) [reply Testing/SystemTests/tests/framework/SANSILLAutoProcessTest.py:205]: The first and the last have the same wavelength of 4.6, and the middles is at 9.99, according to the meta-data.

**Comment by darominski** [comment]
> Can you make the example in PR description to actually use 2 different transmissions for the same sample?

This is now changed to use the SystemTest where there are fewer samples but each sample has its own transmission.

File: docs/source/release/v6.0.0/sans.rst [hunk 3]
```diff
@@ -27,7 +27,12 @@ Improvements
  - With :ref:`SANSILLAutoProcess <algm-SANSILLAutoProcess>`, if sample thickness is set to -1, the algorithm will try to get it
    from the nexus file.
  - With :ref:`SANSILLAutoProcess <algm-SANSILLAutoProcess>`, the output workspace will get its title from the nexus file.
+ - In :ref:`SANSILLAutoProcess <algm-SANSILLAutoProcess>`, more than one transmission per processing, consistent with number
```
Comment (gvardany) [inline docs/source/release/v6.0.0/sans.rst:30]: This got duplicated now.

Comment (darominski) [reply docs/source/release/v6.0.0/sans.rst:30]: Removed.

**Comment by gvardany** [comment]
I think there is typo in the script to test. The middle numor for sample is the same as for sample transmission.

**Comment by gvardany** [comment]
The same is for the container, and in the system test, that's why the results don't make sense.

**Comment by gvardany** [comment]
The new example is much better. However there is a warning for container transmission: 
```
Different distances detected! 988: 8.00100040435791, 1020: 38.95899963378906
Different distances detected! 988: 8.00100040435791, 1020: 38.95899963378906
```
Is there not a run with a matching distance and wavelength?
For the sake of nice output, I'd also suggest to create the masks in the test setup itself using MaskBTP; the existing masks are not masking large enough, so the low-q of each curve looks bad.

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLReduction.py [hunk 4]
```diff
@@ -61,6 +61,22 @@ def _check_distances_match(ws1, ws2):
...
 
+    @staticmethod
+    def _check_wavelengths_match(ws1, ws2):
+        """
+            Checks if the wavelength difference between the data is close enough
+            @param ws1 : workspace 1
+            @param ws2 : workspace 2
+        """
+        tolerance = 0.01 # A
+        wavelength_1 = ws1.getRun().getLogData('wavelength').value
```
Comment (gvardany) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLReduction.py:72]: This works except in TOF, where the wavelength in the logs is 0. For TOF, transmissions and fluxes are functions of wavelength, so this check doesn't make sense. So the checks should be performed if it is not in TOF, there must be a flag somewhere in the algorithm already.

Comment (darominski) [reply Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLReduction.py:72]: Noticed and corrected, also added masking of the instrument edges, as requested, and fixed the release notes conflict.

**Comment by gvardany** [comment]
Thanks for the changes, this is good to see light now.

**Review by gvardany — APPROVED** [review]

**Pull request merged**